gh-103134: Update multiprocessing.managers.ListProxy and multiprocessing.managers.DictProxy

invisibleroads opened the pull request:
```
from multiprocessing import Manager
with Manager() as manager:
    xs = manager.list()
    xs.clear()
```

For now, we can use the workaround `del xs[:]`

<!--
Thanks for your contribution!
Please read this comment in its entirety. It's quite important.

# Pull Request title

It should be in the following format:

```
gh-NNNNN: Summary of the changes made
```

Where: gh-NNNNN refers to the GitHub issue number.

Most PRs will require an issue number. Trivial changes, like fixing a typo, do not need an issue.

# Backport Pull Request title

If this is a backport PR (PR made against branches other than `main`),
please ensure that the PR title is in the following format:

```
[X.Y] <title from the original PR> (GH-NNNN)
```

Where: [X.Y] is the branch name, e.g. [3.6].

GH-NNNN refers to the PR number from `main`.

-->


<!-- gh-issue-number: gh-103134 -->
* Issue: gh-103134
<!-- /gh-issue-number -->

invisibleroads committed:
Add support for manager.list().clear()

```
from multiprocessing import Manager
with Manager() as manager:
    xs = manager.list()
    xs.clear()
```

For now, we can use the workaround `del xs[:]`

blurb-it[bot] committed:
📜🤖 Added by blurb_it.

invisibleroads committed:
Update 2023-03-30-18-19-53.gh-issue-103134.bHrn91.rst

terryjreedy commented on Lib/multiprocessing/managers.py:
```diff
@@ -1148,7 +1148,7 @@ def set(self, value):
     '__mul__', '__reversed__', '__rmul__', '__setitem__',
-    'append', 'count', 'extend', 'index', 'insert', 'pop', 'remove',
+    'append', 'clear', 'count', 'extend', 'index', 'insert', 'pop', 'remove',
     'reverse', 'sort', '__imul__'
```
Copy added to list at same time as clear.  See issue.
```suggestion
    'append', 'clear', 'copy', 'count', 'extend', 'index', 'insert', 'pop',
    'remove', 'reverse', 'sort', '__imul__'
```

invisibleroads replied:
Added ListProxy.copy as you suggested

terryjreedy reviewed (commented):
Are there any tests for BaseList Proxy or derivatives that are not automatically expanded by expanding this list?  And which should be manually augmented?

terryjreedy commented on Misc/NEWS.d/next/Library/2023-03-30-18-19-53.gh-issue-103134.bHrn91.rst:
```diff
@@ -0,0 +1 @@
+Add support for ``multiprocessing.Manager().list().clear()``
```
```suggestion
Add support for ``multiprocessing.Manager().list().clear()`` and ``.copy()``.
```

terryjreedy commented on Lib/multiprocessing/managers.py:
```diff
@@ -1148,7 +1148,7 @@ def set(self, value):
     '__mul__', '__reversed__', '__rmul__', '__setitem__',
-    'append', 'count', 'extend', 'index', 'insert', 'pop', 'remove',
+    'append', 'clear', 'count', 'extend', 'index', 'insert', 'pop', 'remove',
     'reverse', 'sort', '__imul__'
```
Given the definition of method `__iadd__` in the definition of ListProxy below, it would seem that it should be in the list above.  On the other hand, adding xs += [2] to your example works as is.  Maybe '___imul__' does not need to be listed.  Does it hurt to list '__iadd__'?

invisibleroads replied:
It looks like `ListProxy.__iadd__` uses `BaseListProxy.extend` and `ListProxy.__imul__` calls `BaseListProxy.__imul__`. I'm not sure of the reasoning behind why `ListProxy` defines `__iadd__` and `__imul__` separately instead of including them directly in `BaseListProxy`. Testing `xs += [2]` and `xs *= 2` works when tested manually.

I'm afraid of replacing `BaseListProxy` with `ListProxy` for fear of breaking something. Looking at `git blame`, it looks like this part of the code was last changed 15 years ago? @benjaminp would it be safe to replace `BaseListProxy` with `ListProxy` and just include `__iadd__` and `__imul__`?

invisibleroads committed:
Update ListProxy and DictProxy

```
from multiprocessing import Manager
with Manager() as manager:
    xs = manager.list()
    xs.clear
    xs.copy
    d = manager.dict()
    d | {}  # __or__
    {} | d  # __ror__
    reversed(d)  # __reversed__
    d.fromkeys
```

suggested by @terryjreedy
tested manually in Python 3.10.8

https://github.com/python/cpython/issues/103134

invisibleroads committed:
Update Misc/NEWS.d/next/Library/2023-03-30-18-19-53.gh-issue-103134.bHrn91.rst

Co-authored-by: Terry Jan Reedy <tjreedy@udel.edu>

invisibleroads committed:
Update 2023-03-30-18-19-53.gh-issue-103134.bHrn91.rst

invisibleroads committed:
Add DictProxy.__ior__ for d |= {'b': 2}

invisibleroads committed:
Update 2023-03-30-18-19-53.gh-issue-103134.bHrn91.rst

invisibleroads committed:
Merge branch 'python:main' into patch-1

invisibleroads committed:
Fix __ior__ and add tests thanks to https://github.com/python/cpython/issues/103134#issuecomment-1493429209 @sunmy2019

invisibleroads committed:
Merge branch 'main' into patch-1

invisibleroads committed:
Merge branch 'main' into patch-1

invisibleroads committed:
Merge branch 'main' into patch-1

invisibleroads committed:
Merge branch 'main' into patch-1

invisibleroads committed:
Merge branch 'main' into patch-1

invisibleroads committed:
Merge branch 'main' into patch-1

invisibleroads committed:
Merge branch 'main' into patch-1

invisibleroads committed:
Merge branch 'main' into patch-1

invisibleroads committed:
Merge branch 'main' into patch-1

invisibleroads committed:
Merge branch 'main' into patch-1

invisibleroads committed:
Merge branch 'main' into patch-1

invisibleroads committed:
Merge branch 'main' into patch-1

invisibleroads committed:
Merge branch 'main' into patch-1

serhiy-storchaka commented on Lib/multiprocessing/managers.py:
```diff
@@ -1146,10 +1146,10 @@ def set(self, value):
-    '__mul__', '__reversed__', '__rmul__', '__setitem__',
-    'append', 'count', 'extend', 'index', 'insert', 'pop', 'remove',
-    'reverse', 'sort', '__imul__'
+    '__add__', '__contains__', '__delitem__', '__getitem__', '__imul__',
```
Why add `__imul__` here?

invisibleroads replied:
I added `__imul__` there because it looks like the attributes are listed in alphabetical order.

serhiy-storchaka commented on Lib/test/_test_multiprocessing.py:
```diff
@@ -5706,12 +5706,16 @@ def _test_list(cls, obj):
         case.assertEqual(obj[0], 5)
         case.assertEqual(obj.count(5), 1)
         case.assertEqual(obj.index(5), 0)
+        obj += [7]
```
Test the type and the value of the list.

Test `*=`.

invisibleroads replied:
Tests added

serhiy-storchaka commented on Lib/test/_test_multiprocessing.py:
```diff
@@ -5706,12 +5706,16 @@ def _test_list(cls, obj):
         case.assertEqual(obj.count(5), 1)
         case.assertEqual(obj.index(5), 0)
+        obj += [7]
+        obj.extend(obj * 2 + obj.copy())
```
Test the type and the value of both `obj * 2` and `obj.copy()`.

invisibleroads replied:
Tests added

serhiy-storchaka commented on Lib/test/_test_multiprocessing.py:
```diff
@@ -5730,7 +5734,17 @@ def _test_dict(cls, obj):
         case.assertListEqual(list(obj.values()), [5])
         case.assertDictEqual(obj.copy(), {'foo': 5})
-        case.assertTupleEqual(obj.popitem(), ('foo', 5))
+        obj |= {'bar': 6}
```
Test the type and the value of `obj`.

invisibleroads replied:
Tests added

serhiy-storchaka commented on Lib/multiprocessing/managers.py:
```diff
@@ -1160,14 +1160,19 @@ def __imul__(self, value):
-    '__setitem__', 'clear', 'copy', 'get', 'items',
-    'keys', 'pop', 'popitem', 'setdefault', 'update', 'values'
+BaseDictProxy = MakeProxyType('DictProxy', (
+    '__contains__', '__delitem__', '__getitem__', '__ior__', '__iter__',
```
Why add `__ior__` here?

invisibleroads replied:
For the same reason above, I wanted to respect the alphabetical order of the attributes.

serhiy-storchaka reviewed (commented):
Please test the type and the value of results.

invisibleroads committed:
Merge branch 'main' into patch-1

invisibleroads committed:
Test the type and the value of the results as requested by @serhiy-storchaka

invisibleroads committed:
Remove trailing whitespace

invisibleroads committed:
Merge branch 'main' into patch-1

invisibleroads committed:
Merge branch 'main' into patch-1

invisibleroads committed:
Merge branch 'main' into patch-1

invisibleroads committed:
Merge branch 'main' into patch-1

invisibleroads committed:
Merge branch 'main' into patch-1

invisibleroads committed:
Merge branch 'main' into patch-1

invisibleroads committed:
Update _test_multiprocessing.py

Check that the dictionary returned by dict_proxy.from_keys is a dict as defined in dict.from_keys as requested by @encukou

encukou committed:
Add links to the NEWS blurb
